net: tc: Limit the max amount of packets in the queue per traffic class

clamattia opened the pull request:
To avoid starvation of a traffic class by another, limit the maximum amount of packets, that can sit in a single traffic-class-fifo to a fraction of the maximum amount of available rx-packets.

Commit:
net: tc: Limit the max amount of packets in the traffic-class-fifo

To avoid starvation of a traffic class by another, limit the maximum amount
of packets, that can sit in a single traffic-class-fifo to a fraction of
the maximum amount of available packets. In the tx-case also reserve
packets for direct sending, in the case, where
CONFIG_NET_TC_SKIP_FOR_HIGH_PRIO is enabled.

Signed-off-by: Cla Mattia Galliard <cla-mattia.galliard@zuehlke.com>

rlubos commented on subsys/net/ip/net_tc.c:
```diff
@@ -52,6 +52,13 @@ bool net_tc_submit_to_tx_queue(uint8_t tc, struct net_pkt *pkt)
 
+	if (k_sem_take(&tx_classes[tc].fifo_slot, K_NO_WAIT) != 0) {
+		net_pkt_unref(pkt);
+		/* TODO update statistics? */
```
It seems to me that `struct net_stats_tc` could have a drop counter, currently there's no such thing.

clamattia replied:
Could you give some guidance? I could add the respective entry in the struct. But what should be done with it in this scenario?

Thanks for the help.

rlubos replied:
If we wanted to keep track of packets dropped here we'd need to:
* Update struct here: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/net_stats.h#L319
* Add function to bump the statistic (similar to https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_stats.h#L584), and call it here
* And preferably update the network shell module to include the new statistic: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/shell/stats.c#L363

Maybe @jukkar could confirm whether all that makes sense

clamattia replied:
I believe, that it may be best, that these function return a status to the caller `verdict` for example and for the caller to then drop the packet and update statistics in `net_core: net_queue_rx` and `net_if: net_if_queue_tx` respectively, it is the function where other stats related stuff happens.

clamattia replied:
It may also fit nicely into the pattern.
`NET_OK` means the packet was enqueued successfully.
`NET_DROP` means, that the packet was not enqueued. The caller is supposed to drop it and update stats.

clamattia replied:
Added

rlubos commented on subsys/net/ip/net_tc.c:
```diff
@@ -52,6 +52,13 @@ bool net_tc_submit_to_tx_queue(uint8_t tc, struct net_pkt *pkt)
 #if NET_TC_TX_COUNT > 0
 	net_pkt_set_tx_stats_tick(pkt, k_cycle_get_32());
 
+	if (k_sem_take(&tx_classes[tc].fifo_slot, K_NO_WAIT) != 0) {
```
Could we hide this behind `#if NET_TC_TX_COUNT > 1`? IMO there's no need to waste cycles on semphore operations that should have no effect if NET_TC_TX_COUNT == 1.

Same comment for the RX part.

rlubos commented on subsys/net/ip/net_tc.c:
```diff
@@ -254,6 +269,8 @@ static void tc_rx_handler(void *p1, void *p2, void *p3)
 			continue;
 		}
 
+		k_sem_give(fifo_slot);
```
Ditto.

clamattia replied:
Adjusted.

clamattia commented on include/zephyr/net/net_if.h:
```diff
@@ -614,6 +614,9 @@ struct net_traffic_class {
 	struct k_fifo fifo;
 
+	/** Semaphore for tracking the available slots in the fifo */
+	struct k_sem fifo_slot;
```
Ok, I would need `#if NET_TC_TX_COUNT  > 1 || NET_TC_RX_COUNT > 1` , I think. Shall I add this?

pdgendt replied:
I see, didn't notice this, @rlubos WDYT?

rlubos replied:
The code ended up ifdef sliced a bit, but on the other hand it makes this change completely neutral to the most common scenario (1 or less traffic class) so IMO it's fine.

clamattia replied:
Agree, code is ugly to safe a bit RAM.
You decide.

clamattia replied:
I reverted this due to another case needed with latest changes
Would now be:
`#if NET_TC_TX_COUNT  > 1 || NET_TC_RX_COUNT > 1 || defined(CONFIG_NET_TC_SKIP_FOR_HIGH_PRIO)`
Let me know if I should add this.

clamattia replied:
@jukkar

jukkar replied:
Please add any relevant `#if ...` to compile out the semaphore if it is not used. The commit seems to be in multiple PRs, I thought I already commented this.

clamattia replied:
Added this back again

pdgendt commented on subsys/net/ip/net_tc.c:
```diff
@@ -317,11 +346,15 @@ void net_tc_tx_init(void)
 
 		k_fifo_init(&tx_classes[i].fifo);
+		max_slots = CONFIG_NET_PKT_TX_COUNT / NET_TC_TX_COUNT;
+		k_sem_init(&rx_classes[i].fifo_slot, max_slots, max_slots);
```
You will need the same wrapping here then.

clamattia replied:
right, thanks

clamattia replied:
done

clamattia commented:
Something that came to mind. The proposed approach is too conservative. It could be allowed for higher priority queues to hold all available packets. It could be only forbidden for lower priority queues.
Proposal as outlined by this pr so far:
```
CONFIG_NET_PKT_RX_COUNT=40
CONFIG_NET_TC_RX_COUNT=4
rx_q[0] max_slots : 10
rx_q[1] max_slots : 10
rx_q[2] max_slots : 10
rx_q[3] max_slots : 10
```
Less conservative, but also starvation avoiding:
 ```
CONFIG_NET_PKT_RX_COUNT=40
CONFIG_NET_TC_RX_COUNT=4
rx_q[0] max_slots : 10
rx_q[1] max_slots : 20
rx_q[2] max_slots : 30
rx_q[3] max_slots : 40
```

clamattia commented:
Fixed copy paste mistake `rx_classes` vs `tx_classes` and removed the wrapper around `k_fifo_put`.

clamattia commented:
- Also avoid starvation of tx, if `CONFIG_NET_TC_SKIP_FOR_HIGH_PRIO` is set. It appears as another pseudo traffic class.
- Removed the conditional addition of `k_sem` in favor of having it always due to added complexity of another case (can add it back if desired.

Commit:
net: tc: Add statistics about dropped packets

Add statistics about packets dropped in net_tc.

Signed-off-by: Cla Mattia Galliard <cla-mattia.galliard@zuehlke.com>

clamattia commented:
- Added statistics in separate commit
- Aligned the `net_pkt_set_rx_stats_tick` call (let me know if this is not correct)
- `Sent` statistics only updated if not dropped
- Remove unconditional increment and then decrement of `tx_pending`. Let me know, if this is not correct.
- Added to statistics shell command. Let me know, if additional tab(s) is needed for nice layout.

Thanks in advance

clamattia commented:
Fixed silly mistakes

clamattia commented:
Rebased and fixed build failure in tests.

clamattia commented:
Fixed minor oversight

jukkar commented on subsys/net/ip/net_core.c:
```diff
@@ -470,21 +468,29 @@ static void net_queue_rx(struct net_if *iface, struct net_pkt *pkt)
 #endif
 
+	net_pkt_set_rx_stats_tick(pkt, k_cycle_get_32());
+
 	if (NET_TC_RX_COUNT == 0) {
 		net_process_rx_packet(pkt);
 	} else {
-		net_tc_submit_to_rx_queue(tc, pkt);
+		if (net_tc_submit_to_rx_queue(tc, pkt) != NET_OK) {
+			goto drop;
+		}
```
This commit seems to be mixing statistics with other changes. Please only do the statistics update in this commit.

clamattia replied:
Fair, I have pushed the other stuff in the initial commit. Let me know if more splitting is needed.

jukkar commented on include/zephyr/net/net_stats.h:
```diff
@@ -328,6 +328,8 @@ struct net_stats_tc {
 #endif
```
Tautology in commit `net: net_tc: ...` subject, it is enough to have `net: tc: ...` here.

clamattia replied:
Fixed

jukkar commented on subsys/net/ip/net_tc.c:
```diff
@@ -318,10 +364,14 @@ void net_tc_tx_init(void)
 
 		k_fifo_init(&tx_classes[i].fifo);
 
+		k_sem_init(&tx_classes[i].fifo_slot, NET_TC_TX_SLOTS, NET_TC_TX_SLOTS);
```
missing `#if NET_TC_RX_COUNT > 1`

jukkar commented on subsys/net/ip/net_tc.c:
```diff
@@ -376,10 +426,14 @@ void net_tc_rx_init(void)
 
 		k_fifo_init(&rx_classes[i].fifo);
 
+		k_sem_init(&rx_classes[i].fifo_slot, NET_TC_RX_SLOTS, NET_TC_RX_SLOTS);
```
ditto

clamattia replied:
Added

jukkar commented on include/zephyr/net/net_if.h:
```diff
@@ -614,6 +614,9 @@ struct net_traffic_class {
 	struct k_fifo fifo;
 
+	/** Semaphore for tracking the available slots in the fifo */
+	struct k_sem fifo_slot;
```
Perhaps we need a separate Kconfig option to enable this. User should be able to select if this is to be used or not.

clamattia replied:
I would not do that. If someone enables `NET_TC_XX_COUNT>1` they expect the different traffic classes not being able to starve each other.

clamattia replied:
Otherwise it is effectively as if `NET_TC_XX_COUNT==1`

jukkar replied:
Hmm, I commented this change already in the other PR. So we need here
```
#if NET_TC_TX_COUNT > 1
```

jukkar reviewed (commented):
Some initial comments for this, will continue the review later.

clamattia commented:
- Fixed suggestions from @jukkar 
- Fixed a problem with `tx_pending`, please review, if I got this right

clamattia commented:
- Fixed `net_pkt_get_len` of possibly unrefed pkt.

clamattia commented:
- rebased
- added `ARG_UNUSED` to all macros, that are used if statistics are disabled. Should they be maded `static inline functions` instead?

clamattia commented:
- fix compliance issue

clamattia commented:
- compliance issue

clamattia commented:
fix more `unused variable` warnings.

clamattia commented:
All tests are passing. Could you have another look @jukkar @rlubos @pdgendt ?
Thank you

rlubos commented:
> fix more unused variable warnings

What were those warnings about? I'm kind of puzzled why would this not be an issue before.

> added ARG_UNUSED to all macros, that are used if statistics are disabled. Should they be maded static inline functions instead?

IMHO static inline functions sound like a better choice that macros in this case.

rlubos reviewed (commented):
Generally looks good to me, just a bit surprised about the complaints about the unused variables out of sudden.

clamattia commented:
> What were those warnings about? I'm kind of puzzled why would this not be an issue before.

In one case the caller was protecting against it with an `ifdef` in another case, the new `len` helper-variable was needed (and conditionally unused), because the packet ownership was already passed to the fifo, when calling the statistics function.

> IMHO static inline functions sound like a better choice that macros in this case.

Shall I update the entire `net_stats.h`? This sounds like something a separate PR should do.

Also:
```C
static inline void f(void *arg) {
#if defined(CONFIG_F)
    /* do it */
#else
    ARG_UNUSED(arg);
#endif
}
```
Or:
```C
#if defined(CONFIG_F)
void f(void *arg) {
    /* do it */
}
#else
static inline void f(void *arg) {
    ARG_UNUSED(arg);
}
#endif
```
?

rlubos commented:
> Shall I update the entire net_stats.h? This sounds like something a separate PR should do.

No need to, just the modified ones.

As for the formatting, either is fine and used in our codebase, but I'd stick to option 2 here to reduce the diff.

clamattia commented:
- rebased
- changed from `macro` to `static inline function` for `net_stats`-shims

rlubos commented on subsys/net/ip/net_stats.h:
```diff
@@ -598,12 +608,62 @@ static inline void net_stats_update_tc_recv_priority(struct net_if *iface,
+	ARG_UNUSED(tc);
+	ARG_UNUSED(priority);
+}
+
```
nit: Redundant empty line, could be fixed if next push is needed

clamattia replied:
Ok, will do if needed.

clamattia replied:
done

clamattia commented:
- rebased
- removed extra newline
- Added conditional compilation for semaphore and simplified the `#if` logic a bit

clamattia commented:
- reverted potential improvement with `net_pkt_set_rx_stats_tick` to be discussed in separate PR to not hold this one off

jukkar reviewed (approved):
@clamattia I am assuming that the same commits in the different PRs are now updated, it looks like that but just wanted to verify it so that there are no surprises when the PRs are merged in different order.

clamattia commented:
> @clamattia I am assuming that the same commits in the different PRs are now updated, it looks like that but just wanted to verify it so that there are no surprises when the PRs are merged in different order.

I tried to do it like this yes.
Safest is, this will merge first and then I can rebase other onto main.
